### PR TITLE
Sample-all efficiency and IndexRecord generation

### PR DESCRIPTION
**Pull request by djtfmartin**
Changes to make sampling all records more efficient, and the introduction of `IndexRecord` generation which is a representation stored in AVRO of the record which will be sent to SOLR. 

This changes the required steps to load a dataset to the following (using `la-pipelines` cmd tool)

1.  `dwca-avro dr123`  => load the dataset from DwCA to verbatim.avro
2. `interpret dr123 `=> interpretation steps (taxonomy, location)
3. `uuid dr123` => generation of globally unique identifiers
4. `image-load dr123` => load images into image service
5. `image-sync dr123` => retrieve image IDs
6. `sample dr123` => sample records
7. `index dr123` => generate index records 
8. `solr dr123` =>  joining samples to index records and push index records to SOLR

To re-sample and re-index all records (with no changes to the data) will involve running:
1. `sample all` => retrieves unique set of lat lngs for all records, and samples
2. `solr all` => joining samples to index records and push index records to SOLR,

Other notes:
* The `ALASampleAvroPipeline` has been removed as this step is no longer used.
* The LatLng pipelines have been consolidated into 1 pipeline for single dataset and all datasets
* The `sample dr123` and `sample all` commandline tools run the exports so the export of lat lng doesn't need to be ran separately.

### CODE REVIEW

File: livingatlas/pipelines/pom.xml [hunk 1]
```diff
@@ -24,6 +24,27 @@
     </testResources>
 
     <plugins>
+<!--      <plugin>-->
```
Comment (timrobertson100) [inline livingatlas/pipelines/pom.xml:27]: Suggest remove properly

File: livingatlas/pipelines/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+import org.apache.solr.common.cloud.Replica;
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.common.cloud.ZkStateReader;
+import org.apache.solr.common.params.CursorMarkParams;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
Comment (timrobertson100) [inline livingatlas/pipelines/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java:73]: Suggest add a comment here explaining clearly why this exists, what version it is based on, and what is different. Alternatively, and perhaps better, would be 2 isolated commits ("#123 Adding SolrIO from Beam 2.17.0", "#123 Patching SolrIO") so it is clear for others what has happened, especially if future changes are made.

File: livingatlas/pipelines/pom.xml (hunk 1 above)
Comment (djtfmartin) [reply livingatlas/pipelines/pom.xml]: done

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+                                    case DOUBLE:
+                                      builder.getDoubles().put(f.name(), (Double) r);
+                                      break;
+                                    case INT:
+                                      builder.getInts().put(f.name(), (Integer) r);
+                                      break;
+                                    case LONG:
+                                      builder.getLongs().put(f.name(), (Long) r);
+                                      break;
+                                    default:
```
Comment (adam-collins) [inline livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java:522]: Any need for `case ARRAY` to `multiValues`?

Does this need to align with `convertIndexRecordToSolrDoc` and `convertSolrDocToIndexRecord`?

File: livingatlas/pipelines/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java (hunk 2 above)
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java]: Ive removed SolrIO override as the problem with timeout was disk usage (100%) on test environment. Test without the extended client timeout have been error free so far.

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/beam/IndexRecordToSolrPipeline.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      ALASolrPipelineOptions options, Pipeline p) {
+    if (options.getDatasetId() == null || "all".equalsIgnoreCase(options.getDatasetId())) {
+      return p.apply(
+          AvroIO.read(IndexRecord.class)
+              .from(
+                  String.join(
+                      "/",
+                      options.getAllDatasetsInputPath(),
+                      "index-record",
+                      options.getDatasetId() + "/*.avro")));
```
[the remark this answers is not in the document]
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/au/org/ala/pipelines/beam/IndexRecordToSolrPipeline.java:240]: Ive fixed this now.

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+    indexRecord.setDoubles(new HashMap<>());
+    indexRecord.setMultiValues(new HashMap<>());
+    List<String> assertions = new ArrayList<String>();
+
+    addToIndexRecord(lr, indexRecord, skipKeys);
+    addToIndexRecord(tr, indexRecord, skipKeys);
+    addToIndexRecord(br, indexRecord, skipKeys);
+    addToIndexRecord(er, indexRecord, skipKeys);
+    addToIndexRecord(mdr, indexRecord, skipKeys);
+    addToIndexRecord(mdr, indexRecord, skipKeys);
```
[the remark this answers is not in the document]
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java]: removed

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/beam/IndexRecordToSolrPipeline.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    return p.apply(
+        AvroIO.read(IndexRecord.class)
+            .from(String.join("/", options.getAllDatasetsInputPath(), "index-record", "*/*.avro")));
```
[the remark this answers is not in the document]
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/au/org/ala/pipelines/beam/IndexRecordToSolrPipeline.java:244]: Yes, you where right. This is fixed now.

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java (hunk 3 above)
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/au/org/ala/pipelines/transforms/IndexRecordTransform.java:525]: Good point. I've aligned them now.

File: livingatlas/pipelines/src/main/java/au/org/ala/pipelines/java/IndexRecordPipeline.java [hunk 7]
```diff
@@ -363,7 +357,7 @@ public static void run(ALASolrPipelineOptions options, ExecutorService executor)
 
     log.info("Joining avro files...");
     // Join all records, convert into string json and IndexRequest for ES
-    Function<BasicRecord, SolrInputDocument> indexRequestFn =
+    Function<BasicRecord, IndexRecord> indexRequestFn =
```
[the remark this answers is not in the document]
Comment (djtfmartin) [reply livingatlas/pipelines/src/main/java/au/org/ala/pipelines/java/IndexRecordPipeline.java:336]: I think ive removed the duplication

**Review by adam-collins — APPROVED** [review]
No issues found.

**Pull request merged**